src/rgw/rgw_sal_daos.cc

```diff
@@ -0,0 +1,2803 @@
+    }
+  }
+  if (ret == 0 || ret == EEXIST) {
+    ret = 0;
+    _is_open = true;
+  } else {
+    ldpp_dout(dpp, 0) << "ERROR: failed to open daos object ("
+                      << get_bucket()->get_name() << ", " << get_key().to_str()
+                      << "): ret=" << ret << dendl;
+  }
+  return ret;
+}
+
+int DaosObject::close(const DoutPrefixProvider* dpp) {
+  if (!is_open()) {
+    return 0;
+  }
+
+  int ret = dfs_release(dfs_obj);
+  ldpp_dout(dpp, 20) << "DEBUG: dfs_release ret=" << ret << dendl;
+
+  if (ret == 0) {
+    _is_open = false;
+  }
+  return ret;
+}
+
+int DaosObject::write(const DoutPrefixProvider* dpp, bufferlist&& data,
+                      uint64_t offset) {
+  d_sg_list_t wsgl;
+  d_iov_t iov;
+  d_iov_set(&iov, data.c_str(), data.length());
+  wsgl.sg_nr = 1;
+  wsgl.sg_iovs = &iov;
+  int ret = dfs_write(get_daos_bucket()->dfs, dfs_obj, &wsgl, offset, nullptr);
+  if (ret != 0) {
+    ldpp_dout(dpp, 0) << "ERROR: failed to write into daos object ("
+                      << get_bucket()->get_name() << ", " << get_key().to_str()
+                      << "): ret=" << ret << dendl;
+  }
+  return ret;
+}
+
+int DaosObject::read(const DoutPrefixProvider* dpp, bufferlist& data,
+                     uint64_t offset, uint64_t& size) {
+  d_iov_t iov;
+  d_iov_set(&iov, data.append_hole(size).c_str(), size);
+
+  d_sg_list_t rsgl;
+  rsgl.sg_nr = 1;
+  rsgl.sg_iovs = &iov;
+  rsgl.sg_nr_out = 1;
+  int ret = dfs_read(get_daos_bucket()->dfs, dfs_obj, &rsgl, offset, &size, nullptr);
+  if (ret != 0) {
+    ldpp_dout(dpp, 0) << "ERROR: failed to read from daos object ("
+                      << get_bucket()->get_name() << ", " << get_key().to_str()
+                      << "): ret=" << ret << dendl;
+  }
+  return ret;
+}
+
+DaosAtomicWriter::DaosAtomicWriter(
+    const DoutPrefixProvider* dpp, optional_yield y,
+    std::unique_ptr<rgw::sal::Object> _head_obj, DaosStore* _store,
+    const rgw_user& _owner, RGWObjectCtx& obj_ctx,
+    const rgw_placement_rule* _ptail_placement_rule, uint64_t _olh_epoch,
+    const std::string& _unique_tag)
+    : Writer(dpp, y),
+      store(_store),
+      owner(_owner),
+      ptail_placement_rule(_ptail_placement_rule),
+      olh_epoch(_olh_epoch),
+      unique_tag(_unique_tag),
+      obj(_store, _head_obj->get_key(), _head_obj->get_bucket()) {}
+
+int DaosAtomicWriter::prepare(optional_yield y) {
+  int ret = obj.open(dpp, DaosObjectOpen::Create);
+  return ret;
+}
+
+// XXX: Do we need to accumulate writes as motr does?
+int DaosAtomicWriter::process(bufferlist&& data, uint64_t offset) {
+  if (data.length() == 0) {
+    return 0;
+  }
+
+  int ret = 0;
+  if (!obj.is_open()) {
+    ret = obj.open(dpp, DaosObjectOpen::Lookup);
+    if (ret != 0) {
+      return ret;
+    }
+  }
+
+  // XXX: Combine multiple streams into one as motr does
+  uint64_t data_size = data.length();
+  ret = obj.write(dpp, std::move(data), offset);
```

Comment:
How are the concurrent writes handled? In case of Rados (& also dbstore), a unique object-ID is created for each object data write to prevent any overwrite & in AtomicWriter::complete() that object-ID is stored in the object metadata to refer to in object::read()

Reply:
We did not handle this before, I will sketch something up to handle this. Adding a TODO for now